src/rcc/backup.rs

```diff
@@ -0,0 +1,111 @@
+//! Backup Power Domain
+
+/// Backup Power Domain
+#[allow(non_snake_case)]
+pub struct Backup {
+    #[cfg(feature = "rtc")]
+    pub RTC: Rtc,
+}
+
+impl Backup {
+    /// Creates a new `Backup`
+    ///
+    /// # Safety
+    ///
+    /// Must only be called once, after the backup domain write
+    /// protection is disabled.
+    pub(crate) unsafe fn new_singleton() -> Self {
+        Self {
+            #[cfg(feature = "rtc")]
+            RTC: Rtc {
+                _marker: core::marker::PhantomData,
+            },
+        }
+    }
+}
+
+#[cfg(feature = "rtc")]
+pub use rtc::{Rtc, RtcClkSel};
+
+#[cfg(feature = "rtc")]
+mod rtc {
+    use crate::rcc::rec::ResetEnable;
+    use crate::stm32::RCC;
+    use core::marker::PhantomData;
+    use cortex_m::interrupt;
+
+    /// Reset, Enable and Clock functionality for RTC
+    pub struct Rtc {
+        pub(super) _marker: PhantomData<*const ()>,
+    }
+
+    unsafe impl Send for Rtc {}
+
+    impl ResetEnable for Rtc {
+        #[inline(always)]
+        fn enable(self) -> Self {
+            // unsafe: Owned exclusive access to this bitfield
+            interrupt::free(|_| {
+                let bdcr = unsafe { &(*RCC::ptr()).bdcr };
+                bdcr.modify(|_, w| w.rtcen().set_bit());
+            });
+            self
+        }
+        #[inline(always)]
+        fn disable(self) -> Self {
+            // unsafe: Owned exclusive access to this bitfield
+            interrupt::free(|_| {
+                let bdcr = unsafe { &(*RCC::ptr()).bdcr };
+                bdcr.modify(|_, w| w.rtcen().clear_bit());
+            });
+            self
+        }
+        #[inline(always)]
+        fn reset(self) -> Self {
+            // unsafe: Owned exclusive access to this bitfield
+            interrupt::free(|_| {
+                let bdcr = unsafe { &(*RCC::ptr()).bdcr };
+                bdcr.modify(|_, w| w.bdrst().set_bit());
+                bdcr.modify(|_, w| w.bdrst().clear_bit());
+            });
+            self
+        }
+    }
+
+    /// RTC kernel clock source selection
+    pub type RtcClkSel = crate::stm32::rcc::bdcr::RTCSEL_A;
+
+    impl Rtc {
+        /// Returns true if the RTC is enabled.
+        pub fn is_enabled(&self) -> bool {
+            // unsafe: Owned exclusive access to this bitfield
+            interrupt::free(|_| {
+                let bdcr = unsafe { &(*RCC::ptr()).bdcr };
+                bdcr.read().rtcen().bit_is_set()
+            })
+        }
+
+        #[inline(always)]
+        /// Modify a kernel clock for this
+        /// peripheral. See RM0433 Section 8.5.8.
+        ///
+        /// # NOTE
+        ///
+        /// This may only be written one time per peripheral reset.
+        /// Check `get_kernel_clk_mux()` to see if the write succeeded.
+        pub fn kernel_clk_mux(&mut self, sel: RtcClkSel) {
```

Comment:
I changed this from `self` to `&mut self` to make `Rtc::handle_lse_css()` only need to take `&mut self`, but it seems safe to me?

Reply:
Ok, that's reasonable (and certainly not "unsafe" from a memory safety point-of-view).

The reason for requiring `self` is it forces the parent to be mutated when the kernel clock is mutated. You could also argue that actually `handle_lse_css()` should also take `self`, and it should return some new type called `RtcFromLsi` or similar. Thinking in terms of a state machine, `Rtc` and `RtcFromLsi` are the states and `handle_lse_css()` is a state transition - this kind of thinking is common in languages like Haskell and Erlang.

So don't think there's a right-or-wrong answer. Requiring `self` does make it more difficult to use, especially if you're thinking imperatively. But it makes the program easier to reason about if you're thinking in terms of states.

Reply:
That makes a lot of sense. I can envison an `Rtc` parameterized by the clock source. The reset situation makes changing between clock states a bit tricky. I'd like to use this more to better understand what kind of interface would be helpful.